Source/AFResult.swift

```diff
@@ -0,0 +1,108 @@
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+public typealias AFResult<T> = Result<T, Error>
+
+// MARK: - Internal APIs
+
+extension AFResult {
+    /// Returns the associated value if the result is a success, `nil` otherwise.
+    var value: Success? {
+        guard case .success(let value) = self else { return nil }
+        return value
+    }
+
+    /// Returns the associated error value if the result is a failure, `nil` otherwise.
+    var error: Error? {
+        guard case .failure(let error) = self else { return nil }
+        return error
+    }
+
+    /// Initializes an `AFResult` from value or error. Returns `.failure` if the error is non-nil, `.success` otherwise.
+    ///
+    /// - Parameters:
+    ///   - value: A value.
+    ///   - error: An `Error`.
+    init(value: Success, error: Failure?) {
+        if let error = error {
+            self = .failure(error)
+        } else {
+            self = .success(value)
+        }
+    }
+
+    /// Evaluates the specified closure when the `AFResult` is a success, passing the unwrapped value as a parameter.
+    ///
+    /// Use the `flatMap` method with a closure that may throw an error. For example:
+    ///
+    ///     let possibleData: AFResult<Data> = .success(Data(...))
+    ///     let possibleObject = possibleData.flatMap {
+    ///         try JSONSerialization.jsonObject(with: $0)
+    ///     }
+    ///
+    /// - parameter transform: A closure that takes the success value of the instance.
+    ///
+    /// - returns: An `AFResult` containing the result of the given closure. If this instance is a failure, returns the
+    ///            same failure.
+    func flatMap<T>(_ transform: (Success) throws -> T) -> AFResult<T> {
+        switch self {
+        case .success(let value):
+            do {
+                return try .success(transform(value))
+            } catch {
+                return .failure(error)
+            }
+        case .failure(let error):
+            return .failure(error)
+        }
+    }
+
+    /// Evaluates the specified closure when the `AFResult` is a failure, passing the unwrapped error as a parameter.
+    ///
+    /// Use the `flatMapError` function with a closure that may throw an error. For example:
+    ///
+    ///     let possibleData: AFResult<Data> = .success(Data(...))
+    ///     let possibleObject = possibleData.flatMapError {
+    ///         try someFailableFunction(taking: $0)
+    ///     }
+    ///
+    /// - Parameter transform: A throwing closure that takes the error of the instance.
+    ///
+    /// - Returns: An `AFResult` instance containing the result of the transform. If this instance is a success, returns
+    ///            the same success.
+    func flatMapError<T: Error>(_ transform: (Failure) throws -> T) -> AFResult<Success> {
```

Comment:
Do we use this internally?

Reply:
Yep, we do in the functional response APIs a couple of times. I figured we'd want to keep the extension rather than duplicate the logic since it's complex.

Reply:
Only usage I can see for `flatMapError` ATM is in the tests.

Reply:
NVM, I had search limited to tests.